[EC 456] Component Library Content Switching Tabs

shane-melton opened the pull request:
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Add support for content switching tabs in addition to routing tabs. This implementation takes a lot of inspiration from the [Angular Material tabs](https://material.angular.io/components/tabs/overview) ([source here](https://github.com/angular/components/tree/main/src/material/tabs)). 

In order to better support both content tabs and routing tabs (similar to Angular Material), the existing routing tab component was renamed to `<bit-tab-nav-bar>` (with `<bit-tab-link>` items) and the content switching tab component is now called `<bit-tab-group>` (with `<bit-tab>` items).

### `<bit-tab-nav-bar>`

The underlying routing tab remains mostly unchanged, with some improvements to ARIA support (`link` role, `aria-disabled`, `aria-current`) and keyboard navigation (arrow keys and spacebar, [per spec](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/tablist_role#keyboard_interactions)).

Example usage:
```
<bit-tab-nav-bar label="Main">
      <bit-tab-link [route]="['active']">Active</bit-tab-link>
      <bit-tab-link [route]="['item-2']">Item 2</bit-tab-link>
      <bit-tab-link [route]="['item-3']">Item 3</bit-tab-link>
      <bit-tab-link [route]="['disable']" [disabled]="true">Disabled</bit-tab-link>
</bit-tab-nav-bar>
```

### `<bit-tab-group>`

The content tab component supports dynamic tabs, label templates, and optional content preservation. ARIA support was also implemented, including `tab` and `tabpanel` roles, `tabIndex`, `aria-selected`, and `aria-labeledby` attributes.

By default, each tab's content is only rendered to the DOM when the tab is selected (unless the `preserveContent` flag is `true`, in which case it is preserved).

Example usage:
```
<bit-tab-group label="Main Content Tabs">
    <bit-tab label="First Tab">First Tab Content</bit-tab>
    <bit-tab label="Second Tab">Second Tab Content</bit-tab>
    <bit-tab>
      <ng-template bit-tab-label>
        <i class="bwi bwi-search tw-pr-1"></i> Template Label
      </ng-template>
      Template Label Content
    </bit-tab>
    <bit-tab [disabled]="true" label="Disabled">
      Disabled Content
    </bit-tab>
</bit-tab-group>
```

## Code changes

### Content Tabs Changes
- **libs/components/src/tabs/tab-group.component.ts:** Update the `bit-tab-group` component to support content switching and template labels.
- **libs/components/src/tabs/tab-item.component.html:** Removed/replaced by the `bit-tab-list-item` directive.
- **libs/components/src/tabs/tab-label.directive.ts:** Directive used for identifying `<ng-template>` tags that will be used as the tab label instead of a plain text label.

- **libs/components/src/tabs/tab.component.ts:** Add the `bit-tab` component defining tabs and their respective content and labels. Supports both text and template labels for tab headers.

### Routing Tabs Changes
- **libs/components/src/tabs/tab-link.component.html:** Add tab nav link component to be used within the `bit-tab-nav-bar` component. Uses an underlying  `<a>` tag and `[routerLink]` and `routerLinkActive` directives.

- **libs/components/src/tabs/tab-nav-bar.component.ts:** Add the `bit-tab-nav-bar` component which handles rendering the tab links and keyboard navigation.

### Shared Changes
- **libs/components/src/tabs/tabs.module.ts:** Update the module to only export the necessary public components/directives. Keeps implementation specific directives internal to avoid polluting public API.
- **libs/components/src/tabs/tab-list-container.directive.ts:** Internal helper directive used for styling the tab header items in both the content and router tab components.
- **libs/components/src/tabs/tab-list-item.directive.ts:** Internal helper directive used for styling and interacting with tab header items.

- **libs/components/src/tabs/tabs.stories.ts:** Update and add additional tab stories to storybook.
## Screenshots

Content Tabs:
![image](https://user-images.githubusercontent.com/8764515/188728191-9b13236f-ccc5-41c0-9a50-167181b384fa.png)

Route Tabs (content is only off center due to storybook story):
![image](https://user-images.githubusercontent.com/8764515/188728500-96e40cdd-56f3-4bb7-bcbb-bad1eee4ce09.png)


## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

shane-melton commented on libs/components/src/tabs/tab-link.component.html:
```diff
@@ -0,0 +1,13 @@
+<a
+  bit-tab-list-item
+  [routerLink]="disabled ? null : route"
```
`[routerLink]` is set to null when disabled to effectively disable the link. [See docs here](https://angular.io/api/router/RouterLink#properties)

shane-melton commented on libs/components/src/tabs/tab.component.ts:
```diff
@@ -0,0 +1,70 @@
+
+import { BIT_TAB, TabLabelDirective } from "./tab-label.directive";
+
+export const BIT_TAB_GROUP = new InjectionToken<any>("BIT_TAB_GROUP");
```
For some more context, these `BIT_*` `InjectionToken<>` are used to help prevent circular dependencies as the components depend on each other.

coroiu commented on libs/components/src/tabs/tab-label.directive.ts:
````diff
@@ -0,0 +1,29 @@
+ * ```
+ */
+@Directive({
+  selector: "[bit-tab-label], [bitTabLabel]",
````
question: why two different selector types?

shane-melton replied:
Good catch, we should only have the one

coroiu commented on libs/components/src/tabs/tab-label.directive.ts:
```diff
@@ -0,0 +1,29 @@
+ */
+@Directive({
+  selector: "[bit-tab-label], [bitTabLabel]",
+  providers: [{ provide: BIT_TAB_LABEL, useExisting: TabLabelDirective }],
```
question: do we need to provide the tab label? I doesn't seem to be used anywere

shane-melton replied:
Good catch, we don't need it

coroiu commented on libs/components/src/tabs/tab-link.component.ts:
```diff
@@ -0,0 +1,50 @@
+  @ViewChild(TabListItemDirective) tabItem: TabListItemDirective;
+  @ViewChild("rla") routerLinkActive: RouterLinkActive;
+
+  @Input()
+  route: string;
+  @Input() disabled = false;
```
nit: fix inconsistent formatting :)

coroiu commented on libs/components/src/tabs/tab.component.ts:
```diff
@@ -0,0 +1,70 @@
+    // user may have manually set `templateLabel` during creation mode, which would then get
+    // clobbered by `undefined` when the query resolves. Also note that we check that the closest
+    // tab matches the current one so that we don't pick up labels from nested tabs.
+    if (value && value.closestTab === this) {
```
suggestion: have you tried modifying the content child decorator `@ContentChild(TabLabelDirective, { descendants: false })` instead of using `closestTab`?

shane-melton replied:
Actually, this also seems to be a requirement for Angular Materials implementation -- theirs is much more complex with animations and other _material_ things compared to ours

coroiu commented on libs/components/src/tabs/tab-label.directive.ts:
```diff
@@ -0,0 +1,29 @@
+import { Directive, Inject, InjectionToken, Optional, TemplateRef } from "@angular/core";
+
+export const BIT_TAB_LABEL = new InjectionToken<TabLabelDirective>("BitTabLabel");
+export const BIT_TAB = new InjectionToken<any>("BIT_TAB");
```
suggestion: use `unknown` instead of any

coroiu commented on libs/components/src/tabs/tab-link.component.ts:
```diff
@@ -0,0 +1,50 @@
+
+import { TabListItemDirective } from "./tab-list-item.directive";
+import { TabNavBarComponent } from "./tab-nav-bar.component";
+
+@Component({
+  selector: "bit-tab-link",
+  templateUrl: "tab-link.component.html",
+})
+export class TabLinkComponent implements FocusableOption, AfterViewInit, OnDestroy {
```
question: why not use an anchor host tag `a[bit-tab-link]`? I suspect we could get rid of `@HostListener("keydown", ["$event"])` if we did that, but I'm sure there are lots of downsides as well. Maybe you could expand a little on the reasoning for this decision?

shane-melton replied:
Using a directive instead of a component may be possible, however it does add a decent amount of complexity and burdens consumers of the directive with following implementation/style details. 

- Tracking the active tab is done via a `routerLInkActive` directive on the templates `<a>` tag. We'd need to ensure the same setup was performed for each `<a>` tag in the `<bit-tab-nav-bar>`.
- Styling is done via the `bitTabListItem` directive. So they'd need to add that to every `<a>` tag.
- ARIA support would need to be duplicated by the consumer of the component or the directive would need to specify those, likely using more `@HostBindings()`.
- The `[routerLink]` value needs to be `null` when the tab is disabled, so that would be left for consumers to follow.

Overall I find using an explicit component/tag to be more consumer friendly here and makes it less likely the implementation/style will break be incorrectly implemented.

coroiu replied:
You make a compelling argument 😄 Maybe it would be more feasible when they complete the proposed `@HostDirective` decorator https://github.com/angular/angular/issues/8785

Small note: I didn't mean changing to a directive, only changing the host tag but keeping it as a component. However I see now that there would be no point using a component since the template would only consist of `<ng-content>` and then we might as well use directives!

**Another question btw:** What happens if you add `tabindex=-1` to the anchor tag? Could that be a way to get rid of the `@HostListener("keydown", ["$event"])`?

shane-melton replied:
Haha thanks, and yeah I stumbled upon that proposal as well and agree that it would make it more feasible to use an anchor tag host once/if able to add additional directives via a `@HostDirective`.

And unfortunately the `tabindex=-1` still doesn't allow the tab to be selected via the spacebar, so I think we may be stuck with the `@HostListener` for now.

coroiu commented on libs/components/src/tabs/tab-label.directive.ts:
```diff
@@ -0,0 +1,29 @@
+import { Directive, Inject, InjectionToken, Optional, TemplateRef } from "@angular/core";
+
+export const BIT_TAB_LABEL = new InjectionToken<TabLabelDirective>("BitTabLabel");
```
suggestion: this injection token uses different name format, change to `BIT_TAB_LABEL`

shane-melton replied:
Removed the InjectionToken as it was not actually necessary.

coroiu commented on libs/components/src/tabs/tab-label.directive.ts:
```diff
@@ -0,0 +1,29 @@
+import { Directive, Inject, InjectionToken, Optional, TemplateRef } from "@angular/core";
+
+export const BIT_TAB_LABEL = new InjectionToken<TabLabelDirective>("BitTabLabel");
+export const BIT_TAB = new InjectionToken<any>("BIT_TAB");
```
question: shouldn't this be placed in `tab.component.ts` ?

shane-melton replied:
Removed the InjectionTokens as they were not actually necessary for our implementation.

coroiu reviewed (changes requested):
All in all good work! I especially like how well the accessibility features work! I've added some questions and suggestions and also have some general things I've noticed

1. **nit:** `any` is used a few places, I think `unknown` is generally preferred as a more type-safe alternative. Maybe you could do a quick search/replace? 

2. **issue:** the tab header is missing a gray container/padding (see figma)

3. **question:** `forwardRef` and your comment about circular dependencies makes me a bit uneasy. I've noticed that the CDK source has the same things. Have you given this any though? Do we know why? Does complexity increase a lot by trying to get rid of the circular dependencies?

4. **suggestion:** In the rest of the CL, directives seem to have `camelCase` selectors while custom tags (usually components) have `kebab-case`. Just go through and tweak the namings a bit to make sure we follow the same convention

coroiu commented on libs/components/src/tabs/tab-group.component.html:
```diff
@@ -1,6 +1,42 @@
+    </ng-template>
+  </button>
+</div>
+<div class="tw-py-5">
```
issue: I think this might need to be changed to `tw-pt-5` otherwise we get a huge empty space at the bottom of dialogs (see example in screenshot below). Either that or we need to modify the dialog padding `tw-pb-8` which is also quite big by itself. @Hinton did you double check that padding with UX?

![image](https://user-images.githubusercontent.com/2285588/188894113-e8a4a2c7-85ea-4cc1-b220-bc2854968ee9.png)

Hinton replied:
Yes, we can't add padding or margin to the bottom.

shane-melton replied:
Switched to `tw-pt-5` to avoid adding extra padding to the bottom of content tabs

Hinton reviewed (commented):
Please ensure we provide attribution if any code is copied from CDK

Hinton commented:
Can we also create two sub-folders, one for nav items and another for content tabs? Will make it easier to separate them.

Hinton commented on libs/components/src/tabs/tab-group.component.ts:
```diff
@@ -1,7 +1,190 @@
+   */
+  keyManager: FocusKeyManager<TabListItemDirective>;
+
+  constructor(private _changeDetectorRef: ChangeDetectorRef) {
```
Change detector is unused

Hinton commented on libs/components/src/tabs/tab-group.component.ts:
```diff
@@ -1,7 +1,190 @@
+    this._groupId = nextId++;
+  }
+
+  _getTabContentId(id: number): string {
```
Is the goal to only use this internally in the component? If so use `protected` instead of underscore, which is supported since Angular 14.

shane-melton replied:
Yep, that's the intention. I did that originally, but my editor was giving warnings that the `protected` method "cannot be resolved when using the AOT compiler." My guess is my editor isn't entirely aware of the change in Angular 14, but it does work appear to be working correctly so I'm good with changing this.

shane-melton commented on libs/components/src/tabs/tab-list-item.directive.ts:
```diff
@@ -0,0 +1,85 @@
+ * and content tabs (button tags)
+ */
+@Directive({ selector: "[bitTabListItem]" })
+export class TabListItemDirective implements FocusableOption {
```
The `FocusKeyManager` used by the group and navbar components require a publicly accessible `focus()` method on the items it controls ARIA support. 

We could possibly make the `elementRef` public on the directive and pass/map that into the `FocusKeyManager` constructor, but exposing the `elementRef` publicly seems less ideal when compared to encapsulating the `focus()` method explicitly.

shane-melton commented:
> All in all good work! I especially like how well the accessibility features work! I've added some questions and suggestions and also have some general things I've noticed
> 
> 1. **nit:** `any` is used a few places, I think `unknown` is generally preferred as a more type-safe alternative. Maybe you could do a quick search/replace?
> 2. **issue:** the tab header is missing a gray container/padding (see figma)
> 3. **question:** `forwardRef` and your comment about circular dependencies makes me a bit uneasy. I've noticed that the CDK source has the same things. Have you given this any though? Do we know why? Does complexity increase a lot by trying to get rid of the circular dependencies?
> 4. **suggestion:** In the rest of the CL, directives seem to have `camelCase` selectors while custom tags (usually components) have `kebab-case`. Just go through and tweak the namings a bit to make sure we follow the same convention

Thanks for the review and suggestions @coroiu!

1. Good call, the couple uses of `any` are now `unknown`.
2. I added an additional wrapper to style the background gray and give that small spacing above the tabs.
3. The circular dependency with the `forwardRef` seems mostly unavoidable here for the tab links and nav bar. The tab nav links need to communicate to the navbar when the active status changes (from the internal `routerLinkActive` directive) via the `updateActiveLink()` method. It's probably possible to keep track of the active route from the nav bar, but that would require subscribing to the active route and matching/checking against all the routes in the tab links which seems more cumbersome. At least when compared to using the `forwardRef` to communicate between the components directly. But I'm open to re-evaluating this if you see any easier/better way.
4. Done, thanks for bringing that up, much more consistent now.

coroiu reviewed (dismissed):
Good work! Sorry for taking so long to approve, I wanted to try it on my own branch to see how it looks in a dialog, and as you can see it works nicely!

![image](https://user-images.githubusercontent.com/2285588/189835686-53ee3453-7925-4723-8036-de52f629635f.png)

There are some paddings to be worked out but I don't think those are an issue for the tab components. I think I'll upload a PR that allows for the removal of the dialog padding. 

Regarding `forwardRef`, you've clearly thought it through and so I'll trust your conclusions! Even if there was a better way I don't think it's worth the effort right now, and I feel it's more important to get this merge! Maybe when someone revisits this in the future they might get some good ideas 🙂

coroiu commented:
Sorry for dismissing my approval but I just noticed something in the figma design when it comes to the paddings

Here is the design for the tab component (which your PR mirrors very nicely)
![image](https://user-images.githubusercontent.com/2285588/189838151-4d50ee72-ec88-47ce-b670-ec82ac838a12.png)

And here is an example of the tabs being used in a dialog
![image](https://user-images.githubusercontent.com/2285588/189838722-cade0a84-fafd-4ebf-bd08-46ce42a00607.png)

As you can see the tabs have a slight padding that matches with the content. I guess this could be added by a wrapper outside of the scope of the tab component, but I'm not sure if that's the best way. Do we need to support both types?

danielleflinn commented:
> do we need to support both cases (with and w/o padidng), or if this is just a miss in figma?
Nope this was a miss in Figma; I'll update to include the padding.

shane-melton commented:
Nice catch @coroiu! Padding is all fixed and matching the updated Figma now

shane-melton commented:
@coroiu It was also decided with @danielleflinn to add content padding to the tabs and adjust the bottom border of tab header to span the entire content area (to match the examples in tabbed dialogs). We'll still need another PR for the dialog component to optionally remove its content padding (e.g. when the content is a tab group) otherwise we'll still get double padding in dialogs.

coroiu reviewed (dismissed):
Awesome work!

Hinton reviewed (commented):
I haven't re-reviewed this, but I believe my concerns were resolved so should be good.

vvolkgang reviewed (approved):
Merge from master dismissed the previous approval, approving again.
